### PR TITLE
[RFC] Initial prototype for handling handle stack setup and argument/return…

### PR DESCRIPTION
**Pull request by vargaz**
… value conversion in icall wrappers.

### CODE REVIEW

File: mcs/class/corlib/corlib.dll.sources [hunk 1]
```diff
@@ -23,6 +23,7 @@ Mono.Globalization.Unicode/SortKeyBuffer.cs
 Mono.Globalization.Unicode/Normalization.cs
 Mono.Globalization.Unicode/NormalizationTableUtil.cs
 Mono/Runtime.cs
+Mono/RuntimeStructs.cs
```
Comment (lambdageek) [inline mcs/class/corlib/corlib.dll.sources]: `Mono/RuntimeStructs.cs` is missing from the PR.

File: mono/metadata/marshal.c [hunk 2]
```diff
@@ -7752,21 +7814,49 @@ mono_marshal_get_native_wrapper (MonoMethod *method, gboolean check_exceptions,
...
-			mono_mb_emit_native_call (mb, csig, piinfo->addr);
+			mono_mb_emit_native_call (mb, call_sig, piinfo->addr);
+		}
+
+		if (uses_handles) {
+			if (MONO_TYPE_IS_REFERENCE (sig->ret)) {
+				// FIXME: Do this properly
+				mono_mb_emit_byte (mb, CEE_CONV_I);
+				mono_mb_emit_byte (mb, CEE_LDIND_REF);
+			}
```
Comment (lambdageek) [inline mono/metadata/marshal.c]: Also out args.

Comment (vargaz) [inline mono/metadata/marshal.c]: This is just a prototype its missing a lot of stuff.

**Comment by lambdageek** [comment]
On the whole, I think it's a good idea to do icall boilerplate automatically.

If I understand the PR, it just wraps all reftype arguments.  Are there any other cases, will we need some way to specify in the icall sig that it need certain arguments handle-ized? Since we're trying to do more marshaling/unmarshaling on the managed side, the other args we're likely to see are pointers to native structs, so we might be okay with just going with some fixed rule.

**Comment by vargaz** [comment]
This PR does all the work which was previously done by the ICALL_ macros in handle.h automatically, so the there is no need to setup/tear down the handle area, set the pending exception, etc. This would make it much simpler to convert icalls to use handles.

**Comment by lambdageek** [comment]
I haven't dug too deeply, but if using handles will be similar to the MonoError project, the painful parts are almost never in the icalls (but the icall handles boilerplate is pretty easy to mess up, so it's good to get rid of it).  The painful parts are when some function deep in the runtime needs to touch several managed objects; when that happens, we need to thread a handle down into that utility from all its call sites.

So this will help us to get started (and we could perhaps make a checked build feature to whitelist non-handleized icalls so that we can keep track of progress) and to make fewer mistakes, but there will still be a lot of work to do to take advantage of it.

**Comment by lambdageek** [comment]
I'm going to pick this up and work on it.

The main thing I plan on doing for a first pass:
1. Assert if the icall signature has stuff we're not dealing with yet (e.g. managed object return values and out args)
2. Add some mechanism to `icall-def.h` to signal that an icall should be wrapped.

**Comment by lambdageek** [comment]
Already merged a versions of this (#3349)

**Pull request closed without merging**